Migrate Aviant changes from 1.13 to 1.15

oystub opened the pull request:
This PR contains custom the modifications from aviant/1.13 branch which we want to keep. 

Details are found at: https://aviant.atlassian.net/wiki/spaces/TECHNICAL/pages/1364361243/WIP+Updrading+from+PX4+1.13.2

bernhardpg committed:
Increase total number of rally points and geofence points to 512

kibidev committed:
Improve Makefile filter function for github config list

The old version would assume that every line containing a comma was a
board config, it is not improved (still not perfect) to only consider
lines between 'config: [' and ']'.

kibidev committed:
workflows: adapt for aviant 1.15 branch

Squash and minor adaptation of cd8b050, e9a204e and 414f987 from the 1.13 branch

Remove workflwes and steps/builds not relevant. Remove uploads to
services that we do not use. Adapt for branch aviant/1.15 instead of
master and use for PR for all branch names (would not run for branches
with "/" in the name).

Only build default version in workflow
The workflow would build cubepilot_cubeorange_test and _bootloader, that
we do not use. And cubepilot_cubeorange_test is using 100% flash, and
will fail.

workflow: releases built on tag push will be pre-release
This is normally what we want, and we will manually update it to
"latest" when it is ready for production.

kibidev committed:
boards: increase log buffer for cube orange

kibidev committed:
boards: increase Nuttx send buffer for /dev/ttyS0 (USART2)

When sending much data (e.g. streaming log), Mavlink::get_free_tx_buf()
in src/modules/mavlink/mavlink_main.cpp will report too little space in
TX buffer a lot, even though the average send speed is well below max
rate. By increase the buffer, we can handle a spikes in traffic without
dropping messages. The limit in lab testing seem to be somewhere around
6000, so we set a bit higher than that to have some margins.

kibidev committed:
platforms/nuttx: remove i2c general call reset

The general call reset (0x06 command to address 0x00) seem to send
LW20/C into a garble mode that blocks any subsequent  i2c communication.
This might have a FW fix for the sensor that we can upgrade to later,
but for now this is a quick fix.

kibidev committed:
Return to center after loiter to altitude

This commit revert most of
https://github.com/PX4/PX4-Autopilot/pull/18996
so that we will no longer skip ahead to next waypoint if we are ahead of
the current. This will enfource that we return to the center of the
loiter before continue after a loiter to altitude. This is to prevent
that we turn out of the loiter when done, and reach a distance from the
WP/path that is larger than intended.

kibidev committed:
commander: parachute on disarm in air, except right after takeoff

Use COM_LKDOWN_TKO when a disarm while in the air is received to decide
if the parachute should be deployed.

This is a port of b60e37f and ea8fe9d from aviant/1.13

kibidev committed:
uavcan: add parameter for logging verbosity

Add UAVCAN_LOG_LEVEL to control logging for messages received on uavcan.
This enables difference verbosity for uavcan messages and other
messages. The verbosity for uavcan messages can not be higher than
general logging verbosity.

kibidev committed:
mavlink: add ATTITUDE and SYSTEM_TIME to custom mavlink mode

For use with custom AVSS parachute. The ATTITUDE messages are used as a
"fast heartbeat", and SYSTEM_TIME is used to set time. Rate for
SYSTEM_TIME is set to 1 Hz, as this is tested to be enough.

Squash of commits 08ceda9, 63f351c, and d954051 from aviant/1.13

Torjusba committed:
ekf2: Redefine ECL_ERR from PX4_DEBUG to PX4_INFO

Because we want to log these events in non-debug builds without alerting
the pilot

kibidev committed:
vtol_att_control: publish vtol_transition_failsafe and vtol_in_rw_mode in same update

Previously, vtol_transition_failsafe would be published in one iteration
of the loop in vtol_att_control, while vtol_in_rw_mode based on
vtol_transition_failsafe would be updated in the next. This could cause
RTL to create a mission item for MR while the navigator still uses FW
mode for calculating acceptance radius using in the RTL mission item.

Igor-Misic committed:
vtol_att_control: allow override failsafe mode with external commands

bresch committed:
ekf2_params: reduce "short" description

Torjusba committed:
workflow: actions/upload_artifact from v2 to v4

oystub committed:
control_allocator: always prioritize thrust over yaw when saturated

Cherry picked from 89b1566, but without the mixer part, as this is removed since 1.14

kibidev committed:
mixer_module: dynamically set max actuator value based on battery

kibidev committed:
Disable time stretch while landing in auto

If wind is too strong for the drone to hold position, time stretch can
prevent the drone from ever landing. Also reset trajectory setpoint XY
position when starting landing.

This is a re-implementation of commit
3070276 from aviant/1.13
which again comes from
0b93725 from aviant/1.12

Co-authored-by: oystub <o_github@oystub.com>

dakejahl committed:
logger: always log can_interface_status (#24071)

* logger: always log can_interface_status

* logger: log topic can_interface_status using CONFIG_BOARD_UAVCAN_INTERFACES

kibidev committed:
workflow: upgrade actions/cache to v4

oystub committed:
navigator: publish winch commands to MAV_COMP_ID_WINCH

Torjusba committed:
battery: add support for NTC battery thermistor in ADC port

Off by default, enable using the parameter BAT{i}_T_CHANNEL

(cherry picked from commit d97fa0335ddfb7ff396730732607d11d3cf3bed5)
Then added temperature to the battery simulator

haumarco committed:
use imu-timestamp instead of hrt_abstime during replay for aid-srcs (#25117)

Torjusba committed:
ms4525do: log corupted samples

For debugging airspeed sensor errors.

This logs the number of corrupted samples as a perf counter, as well as
the number of samples with corrupted LSBs which have been accepted.

oystub committed:
sensors: add vehicle_gnss_heading to sensor bridge

1.15 port of 8901aba

Modelled after vehicle_gps_position, it processes incoming sensor_gnss_relative
messages and outputs vehicle_gps_yaw with a validated body-frame heading
measurement.

In the future, this can be used in the EKF, instead of the existing hacky
solution which only works with blending enabled.

Note that only one instance is supported. This is because PX4 only has support
for two GNSS receivers anyway, so there isn't "room" for multiple rovers that
output heading.

oystub committed:
ekf2: use vehicle_gnss_heading as source for rtk heading measurements

oystub committed:
vehicle_gps_position: remove blending, use simpler selector logic

This makes it way easier to show that our GNSS fallback logic matches
our compliance documentation.

oystub committed:
gps: option for normal mode fallback if ubx mb rover is used for position

A receiver in moving base rover mode will output data with delay up to 1 second
if it cannot compute a relative solution. This can for example happen if the
moving base experiences degrated signal or power loss.

In such a scenario, the rover might be the only receiver left, and for
redundancy it's important to bring it back into a normal state without delay in
order to avoid toilet bowl/position loss.

(1.15 port of 7a20518)

Torjusba committed:
navigator: add option to ignore a range of icao addresses

For traffic avoidance, this adds the option to give a range of icao
addresses that will not trigger the configured traffic avoidance action.
Can be used to ignore an external transponder in the vehicle, or other
vehicles in the same fleet.

This implements the same functionality for 1.15 as
9aa128d99dd7a81cb98a410e3bc91b2685eedd2c did for 1.13

Torjusba committed:
commander: add failure detection for MPC Z rate
A descent velocity above a configurable limit, while the rate setpoint
is negative (ascending) will trigger failure detection.

This commit implements the same functionality for v1.15 as
7a6a79880eb00906dc06432730a53e35ffc750f2 did for v1.13

oystub committed:
vtol_att_control: filter airspeed measurements in transition logic

Airspeed measurements are filtered using a 1st order IIR low-pass
filter which accounts for sample rate.The filter time constant is
controlled using parameter VT_ARSP_TAU.

This solves an issue where spikes in airspeed values would cause
oscillations in blending weights and in turn strong throttle
oscillations, impairing transition performance.

In addition, the mr/fw blending weights are published to
vtol_vehicle_status for ease of debugging

oystub committed:
Revert "mavlink: add ATTITUDE and SYSTEM_TIME to custom mavlink mode"

This reverts commit ebf348759e7f7e9241c0618c9527fe9d74e6af0c.

oystub committed:
mavlink: add parachute mavlink mode

For use with custom AVSS parachute. The ATTITUDE messages are used as a
"fast heartbeat", and SYSTEM_TIME is used to set time. Rate for
SYSTEM_TIME is set to 1 Hz, as this is tested to be enough.

oystub committed:
Revert "uavcan: add parameter for logging verbosity"

This reverts commit c767df6ef73c4e4333728a10a99dd1438a9c666f.

oystub committed:
adsb: hide "traffic resolved" warnings

These messages are of no relevance to us the way we use traffic warnings.

QGC doesn't show "Debug" level warnings, so this is the same as hiding them.

Torjusba committed:
mission RTL flight mode

Breaking out Mission RTL in a separate flight mode in order to separate
it from rallypoint RTL in failsafe actions, and be able to failsafe from
mission RTL to rallypoint RTL.

Torjusba committed:
convenience script for running SITL in docker

Torjusba committed:
failsafe: remove "Autopilot disengaged" from generic event message

It's not really accurate, since the generic cause can be a lot of
things.
This phrase will likely be more confusing than useful for our pilots.

Torjusba committed:
failsafe: treat RTL flight mode as a failsafe action

In order to get the same override behaviour when the flight mode is
selected manually as with a failsafe.

oystub committed:
cubeorange:  MAVLink on TELEM2 in case of parameter reset

oystub committed:
mavlink: use aviant dialect

oystub committed:
mpc: minimum land speed for battery emergency

In a battery emergency, it may be desirable to limit altitude-based
reductions to vertical speed during landing. A hard landing might be
preferable to running out of battery.

This especially applies if MPC_LAND_CRWL is enabled, since distance
sensors can measure too low distance in foggy conditions, which would
potentially cause a very slow landing, draining the battery.

oystub committed:
aviant: navigation indicator

oystub committed:
ekf2: add vehicle_gnss_heading to replay

oystub committed:
aviant/nav: fix ekf2 estimator nav checks

Torjusba commented:
Force-pushed now, just dropped the one commit for minimum alt agl for flight termination:
https://github.com/aviant-tech/PX4-Autopilot/pull/120
https://github.com/aviant-tech/PX4-Autopilot/pull/91/commits/e6f1c6c72db686aa2a3215676292af28f33da1ce

oystub committed:
ekf2: run simplified GNSS checks after initial fix

This prevents stopping GNSS fusion on slightly degraded solution when
tight checks are set

Backport of functionality from upstream commit
457ce90541204b4ef7f89b3317f6fb2a46b187cb
by bresch <brescianimathieu@gmail.com>

Co-authored-by: bresch <brescianimathieu@gmail.com>

oystub committed:
ekf2: reduce reset intervals

We want the ekf to reset before declaring the local position invalid.

oystub committed:
vehicle_gps_position: run same health checks as ekf

to always prefer the GPS that would be healthy

oystub committed:
ekf2: significantly reduce hysteresis for passing gnss checks

It is very important that we switch back to gnss quickly after recovery

oystub committed:
ekf2: tune the gnss check recovery time

Tests show that it takes some time for height estimate to be good
after swithcing gnss

hermankolden committed:
vtol_att_control: add lookahead option to attitude quadchute

oystub committed:
gnss_selector: use stricter health thresholds

oystub committed:
gnss_selector: always respect hysteresis

hermankolden committed:
vtol_att_control: filter lookahead angles

We used to get too many false positives

hermankolden committed:
vtol_att_control: dont publish mc weights

No code subscribes to published mc weights. It can be useful for
transition troubleshooting, but then it's sufficient with airspeed_filtered.
Removing to simplify code and to halven message size.

Igor-Misic committed:
navigator: add wait flag to VEHICLE_CMD_DO_REPOSITION

The flag ensures that reposition_triplet is not reset after a second VEHICLE_CMD_DO_REPOSITION arrives.

Torjusba committed:
commander: don't force lockdown from flighttermination

This causes the PWM outputs to be disabled, which breaks PWM parachute
triggering

Torjusba reviewed (approved):
This PR is in a good place now, and we are ready to merge it. Process doc https://aviant.atlassian.net/wiki/x/AQBPbw
